# read wiki page to dataframe

In [191]:
import pandas as pd
from geopy.geocoders import Nominatim
import warnings
warnings.filterwarnings("ignore")
import requests
import numpy as np
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

In [192]:
tables=pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
df=tables[0]

df.shape
#df.head()

(180, 3)

### Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

In [157]:
df=df[df["Borough"]!="Not assigned"]
df.shape

(103, 3)

### If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [158]:
# there is no neighborhood as "Not assigned" after removing "Not assigned" Borough

list=df[df["Neighbourhood"] == 'Not assigned']

len(list)


0

### After data clean, there are 103 rows left in table.

In [159]:
df.shape

(103, 3)

In [160]:
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

# Add geo info of postal code

In [161]:
df_geo=pd.read_csv("C:/Users/jingh/Downloads/Geospatial_Coordinates.csv")

In [162]:
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [163]:
df_merge = pd.merge(df, df_geo, on='Postal Code')

In [164]:
df_merge.head(5)

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

# Explore and cluster the neighborhoods in Toronto

### Data set explore

In [165]:
#!conda install -c conda-forge folium=0.5.0 --yes 
import folium 

In [166]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [167]:
# filter out neighborhoods in Toronto and rename Neighbourhood to Neighborhood
df_Toronto=df_merge[df_merge['Borough'].str.contains('Toronto')]
df_Toronto.rename(columns={'Neighbourhood':'Neighborhood'},inplace = True)

In [168]:
df_Toronto.reset_index(drop=True)

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

#### Neighbor distribution in Toronto

In [169]:
# create map of Toronto using latitude and longitude values

map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_Toronto['Latitude'], df_Toronto['Longitude'], df_Toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

### Define Foursquare Credentials and Version

In [170]:
CLIENT_ID = 'YFYZIO4DCIHU2MIYEAZ0QP05CKWF4HA3H5KZLXTAGZUQHR3J' # your Foursquare ID
CLIENT_SECRET = 'JIPMNLBLVO1TLUUOOXYV5SKVRU3VXAZ3B1IBVZISCEGB342L' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 
radius = 500

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: YFYZIO4DCIHU2MIYEAZ0QP05CKWF4HA3H5KZLXTAGZUQHR3J
CLIENT_SECRET:JIPMNLBLVO1TLUUOOXYV5SKVRU3VXAZ3B1IBVZISCEGB342L


### Explore Neighborhoods in Toronto

In [171]:
def getNearbyVenues(borough,names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for bor,name, lat, lng in zip(borough,names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            bor,
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Borough',
                             'Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [172]:
# run the above function on each neighborhood and create a new dataframe called Toronto_venues.
Toronto_venues = getNearbyVenues(borough=df_Toronto['Borough'],
                                 names=df_Toronto['Neighborhood'],
                                   latitudes=df_Toronto['Latitude'],
                                   longitudes=df_Toronto['Longitude']
                                   
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [173]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1647, 8)


Borough               Neighborhood  Neighborhood Latitude  \
0  Downtown Toronto  Regent Park, Harbourfront               43.65426   
1  Downtown Toronto  Regent Park, Harbourfront               43.65426   
2  Downtown Toronto  Regent Park, Harbourfront               43.65426   
3  Downtown Toronto  Regent Park, Harbourfront               43.65426   
4  Downtown Toronto  Regent Park, Harbourfront               43.65426   

   Neighborhood Longitude                   Venue  Venue Latitude  \
0              -79.360636        Roselle Desserts       43.653447   
1              -79.360636           Tandem Coffee       43.653559   
2              -79.360636  Cooper Koo Family YMCA       43.653249   
3              -79.360636     Body Blitz Spa East       43.654735   
4              -79.360636          Impact Kitchen       43.656369   

   Venue Longitude       Venue Category  
0       -79.362017               Bakery  
1       -79.361809          Coffee Shop  
2       -79.358008  Distribution Center  
3       -79.359874                  Spa  
4       -79.356980           Restaurant

In [174]:
Toronto_venues.groupby("Venue Category").count().sort_values(by="Neighborhood",ascending=False)

Borough  Neighborhood  Neighborhood Latitude  \
Venue Category                                                          
Coffee Shop                  155           155                    155   
Café                          92            92                     92   
Restaurant                    54            54                     54   
Italian Restaurant            38            38                     38   
Hotel                         38            38                     38   
...                          ...           ...                    ...   
Garden Center                  1             1                      1   
Fruit & Vegetable Store        1             1                      1   
Frozen Yogurt Shop             1             1                      1   
Flea Market                    1             1                      1   
Harbor / Marina                1             1                      1   

                         Neighborhood Longitude  Venue  Venue Latitude  \
Venue Category                                                           
Coffee Shop                                 155    155             155   
Café                                         92     92              92   
Restaurant                                   54     54              54   
Italian Restaurant                           38     38              38   
Hotel                                        38     38              38   
...                                         ...    ...             ...   
Garden Center                                 1      1               1   
Fruit & Vegetable Store                       1      1               1   
Frozen Yogurt Shop                            1      1               1   
Flea Market                                   1      1               1   
Harbor / Marina                               1      1               1   

                         Venue Longitude  
Venue Category                            
Coffee Shop                          155  
Café                                  92  
Restaurant                            54  
Italian Restaurant                    38  
Hotel                                 38  
...                                  ...  
Garden Center                          1  
Fruit & Vegetable Store                1  
Frozen Yogurt Shop                     1  
Flea Market                            1  
Harbor / Marina                        1  

[234 rows x 7 columns]

### filter out coffee shop

In [175]:
Toronto_coffee=Toronto_venues[Toronto_venues['Venue Category']=="Coffee Shop"]
Toronto_coffee.drop("Venue Category",axis=1,inplace=True)
Toronto_coffee.shape

(155, 7)

### Check how many coffee shop brands

In [176]:
Toronto_coffee['Venue'].value_counts()

Starbucks                            53
Tim Hortons                          25
Pilot Coffee Roasters                10
Dineen @CommerceCourt                 5
Balzac's Coffee                       4
M Square Coffee Co                    4
Everyday Gourmet (Teas & Coffees)     3
Second Cup                            3
Bulldog On The Block                  3
Dark Horse Espresso Bar               2
Versus Coffee                         2
Jimmy's Coffee                        2
Hailed Coffee                         2
Neo Coffee Bar                        2
Aroma Espresso Bar                    2
Mos Mos                               2
Cafe Frappe                           1
Tandem Coffee                         1
Tokyo Smoke                           1
I Deal Coffee                         1
Rooster Coffee House                  1
Arvo                                  1
Balzac’s Coffee Roasters              1
Little Pebbles                        1
Sumach Espresso                       1


In [177]:
# We want to avoid to open my coffee shop near Starbucks
Toronto_target=Toronto_coffee[Toronto_coffee['Venue']!="Starbucks"]
Toronto_target.shape

(102, 7)

In [178]:
Toronto_target['Borough'].describe()

count                  102
unique                   4
top       Downtown Toronto
freq                    84
Name: Borough, dtype: object

### Cluster Neighborhoods

In [179]:
# Run k-means to cluster the existing coffee shop into 5 clusters.
kclusters = 5

Toronto_grouped_clustering = Toronto_target.drop(['Borough','Neighborhood','Neighborhood Latitude','Neighborhood Longitude','Venue'], 1)

Toronto_grouped_clustering

 

Venue Latitude  Venue Longitude
1          43.653559       -79.361809
13         43.658135       -79.359515
14         43.651900       -79.365609
16         43.649963       -79.361442
35         43.653081       -79.357078
...              ...              ...
1544       43.644373       -79.383065
1576       43.668286       -79.382520
1593       43.669654       -79.379871
1620       43.664167       -79.380149
1625       43.669575       -79.382331

[102 rows x 2 columns]

In [180]:
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([4, 4, 4, 4, 4, 4, 1, 0, 1, 1])

In [181]:
Toronto_target.insert(6,'Cluster Labels', kmeans.labels_)
Toronto_target.reset_index(drop=True)

Borough                            Neighborhood  \
0    Downtown Toronto               Regent Park, Harbourfront   
1    Downtown Toronto               Regent Park, Harbourfront   
2    Downtown Toronto               Regent Park, Harbourfront   
3    Downtown Toronto               Regent Park, Harbourfront   
4    Downtown Toronto               Regent Park, Harbourfront   
..                ...                                     ...   
97   Downtown Toronto  First Canadian Place, Underground city   
98   Downtown Toronto                    Church and Wellesley   
99   Downtown Toronto                    Church and Wellesley   
100  Downtown Toronto                    Church and Wellesley   
101  Downtown Toronto                    Church and Wellesley   

     Neighborhood Latitude  Neighborhood Longitude                    Venue  \
0                43.654260              -79.360636            Tandem Coffee   
1                43.654260              -79.360636          Sumach Espresso   
2                43.654260              -79.360636           Rooster Coffee   
3                43.654260              -79.360636                     Arvo   
4                43.654260              -79.360636  Dark Horse Espresso Bar   
..                     ...                     ...                      ...   
97               43.648429              -79.382280          Balzac's Coffee   
98               43.665860              -79.383160      Piedmont Coffee Bar   
99               43.665860              -79.383160     Rooster Coffee House   
100              43.665860              -79.383160                The Drink   
101              43.665860              -79.383160           Croissant Tree   

     Venue Latitude  Cluster Labels  Venue Longitude  
0         43.653559               4       -79.361809  
1         43.658135               4       -79.359515  
2         43.651900               4       -79.365609  
3         43.649963               4       -79.361442  
4         43.653081               4       -79.357078  
..              ...             ...              ...  
97        43.644373               1       -79.383065  
98        43.668286               1       -79.382520  
99        43.669654               1       -79.379871  
100       43.664167               1       -79.380149  
101       43.669575               1       -79.382331  

[102 rows x 8 columns]

In [182]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_target['Venue Latitude'], Toronto_target['Venue Longitude'],Toronto_target['Venue'], Toronto_target['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Clusters

#### Cluster 1:

In [183]:
Cluster1=Toronto_target.loc[Toronto_target['Cluster Labels'] == 0]
Cluster1.shape

(11, 8)

#### Cluster 2:

In [184]:
Cluster2=Toronto_target.loc[Toronto_target['Cluster Labels'] == 1]
Cluster2.shape

(70, 8)

#### Cluster 3: 

In [185]:
Cluster3=Toronto_target.loc[Toronto_target['Cluster Labels'] == 2]
Cluster3.shape

(4, 8)

#### Cluster 4:

In [186]:
Cluster4=Toronto_target.loc[Toronto_target['Cluster Labels'] == 3]
Cluster4.shape

(3, 8)

#### Cluster 5

In [187]:
Cluster5=Toronto_target.loc[Toronto_target['Cluster Labels'] == 4]
Cluster5 .shape

(14, 8)

### Find out the Neighborhood of cluster1

In [188]:
Cluster2['Borough'].unique()

array(['Downtown Toronto'], dtype=object)

In [189]:
Cluster2['Neighborhood'].unique()

array(["Queen's Park, Ontario Provincial Government",
       'Garden District, Ryerson', 'St. James Town', 'Berczy Park',
       'Central Bay Street', 'Richmond, Adelaide, King',
       'Harbourfront East, Union Station, Toronto Islands',
       'Toronto Dominion Centre, Design Exchange',
       'Commerce Court, Victoria Hotel',
       'CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport',
       'Stn A PO Boxes', 'First Canadian Place, Underground city',
       'Church and Wellesley'], dtype=object)

In [190]:
Cluster2.groupby("Neighborhood").count().sort_values(by="Venue",ascending=False)

Borough  \
Neighborhood                                                  
Central Bay Street                                        8   
Commerce Court, Victoria Hotel                            8   
First Canadian Place, Underground city                    8   
Harbourfront East, Union Station, Toronto Islands         7   
Toronto Dominion Centre, Design Exchange                  7   
Stn A PO Boxes                                            6   
Garden District, Ryerson                                  5   
Richmond, Adelaide, King                                  5   
St. James Town                                            5   
Church and Wellesley                                      4   
Berczy Park                                               3   
Queen's Park, Ontario Provincial Government               3   
CN Tower, King and Spadina, Railway Lands, Harb...        1   

                                                    Neighborhood Latitude  \
Neighborhood                                                                
Central Bay Street                                                      8   
Commerce Court, Victoria Hotel                                          8   
First Canadian Place, Underground city                                  8   
Harbourfront East, Union Station, Toronto Islands                       7   
Toronto Dominion Centre, Design Exchange                                7   
Stn A PO Boxes                                                          6   
Garden District, Ryerson                                                5   
Richmond, Adelaide, King                                                5   
St. James Town                                                          5   
Church and Wellesley                                                    4   
Berczy Park                                                             3   
Queen's Park, Ontario Provincial Government                             3   
CN Tower, King and Spadina, Railway Lands, Harb...                      1   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Central Bay Street                                                       8   
Commerce Court, Victoria Hotel                                           8   
First Canadian Place, Underground city                                   8   
Harbourfront East, Union Station, Toronto Islands                        7   
Toronto Dominion Centre, Design Exchange                                 7   
Stn A PO Boxes                                                           6   
Garden District, Ryerson                                                 5   
Richmond, Adelaide, King                                                 5   
St. James Town                                                           5   
Church and Wellesley                                                     4   
Berczy Park                                                              3   
Queen's Park, Ontario Provincial Government                              3   
CN Tower, King and Spadina, Railway Lands, Harb...                       1   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Central Bay Street                                      8               8   
Commerce Court, Victoria Hotel                          8               8   
First Canadian Place, Underground city                  8               8   
Harbourfront East, Union Station, Toronto Islands       7               7   
Toronto Dominion Centre, Design Exchange                7               7   
Stn A PO Boxes                                          6               6   
Garden District, Ryerson                                5               5   
Richmond, Adelaide, King                                5               5   
St. James